File: src/learning_topic/scripts/velocity_publisher.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# 该例程将发布一个turtle/cmd_vel话题，消息类型为geometry_msgs::Twist

import rospy
from geometry_msgs.msg import Twist


def shutdown_hook():
    rospy.loginfo("用户中断退出")

def velocity_publisher():
    # ROS节点初始化
    rospy.init_node('velocity_publisher', anonymous=True)
    #rospy.init_node(name, anonymous=False):name:节点名称，anonymous:不记名的，这样可以运行很多不关心名称的节点

    # 创建一个Publisher，发布名为/turtle1/cmd_vel的topic，消息类型为geometry_msg::Twist，队列长度10
    turtle_vel_pub = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)
    #pub = rospy.Publisher('topic_name', std_msgs.msg.String, queue_size=10) 主题名称，Message类，队列大小

    # 设置循环的频率
    rate = rospy.Rate(10)  # rospy 提供了一个方便的类，该类尽力保持循环的特定速率。

    while not rospy.is_shutdown():
    #ROS 中测试关闭的最常见用法模式是：
    #spin() 代码简单地休眠，直到 is_shutdown() 标志被 True 。它主要用于防止你的 Python 主线程退出。
        # 初始化geometry_msgs::Twist类型的消息
        vel_msg = Twist()
        vel_msg.linear.x = 0.5
        vel_msg.angular.z = 0.2

        # 发布消息
        turtle_vel_pub.publish(vel_msg)
        rospy.loginfo("Publish turtle velocity command[%0.2f m/s, %0.2f rad/s]",
                   vel_msg.linear.x, vel_msg.angular.z)  # 写入日志消息
        # rospy.loginfo("已经修改过代码")

        # 按照循环频率延时
        rate.sleep()  # rospy.sleep(duration) 会睡眠指定的时间



if __name__ == '__main__':
    try:
        rospy.on_shutdown(shutdown_hook)  # 在节点关闭时调用shutdown_hook
        velocity_publisher()
    except rospy.ROSInterruptException:  # ROSInterruptException:用于中断操作的异常
        pass

File: src/learning_topic/src/velocity_publisher.cpp
/**
 * 该历程将发布turtle/cmd_vel话题，消息类型geometry_msgs::Twist
 */

#include <ros/ros.h>                // 包含ROS库头文件
#include <geometry_msgs/Twist.h>    // 包含geometry_msgs::Twist消息类型

int main(int argc, char *argv[])
// argc:整数,用来统计你运行程序时送给main函数的命令行参数的个数
// * argv[ ]: 指针数组，用来存放指向你的字符串参数的指针，每一个元素指向一个参数
{
    // ROS节点初始化
    ros::init(argc, argv, "velocity_publisher");
    //ros::init(argc, argv, "my_node_name", ros::init_options::AnonymousName);
    //node_name:将分配的节点名称；opotions:这是一个可选参数，允许您指定某些选项来改变 roscpp 的行为。

    // 创建节点句柄
    ros::NodeHandle n;

    // 创建一个Publisher，发布为/turtle1/cmd_vel的topic，消息类型为geometry_msgs::Twist，队伍长度10
    ros::Publisher turtle_vel_pub = n.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
    //ros::Publisher advertise(const std::string& topic, uint32_t queue_size, bool latch = false);
    //topic:要发布的主题；queue_size:出站消息的大小

    // 设置循环的频率
    ros::Rate loop_rate(10);  // ros::Rate:该类尽力维持一个特定的循环速率

    int count = 0;  // 计数器，暂时未使用

    while(ros::ok())
    {
        // 初始化geometry_msgs::Twist类型的消息
        geometry_msgs::Twist vel_msg;
        vel_msg.linear.x = 0.5;
        vel_msg.angular.z = 0.2;

        // 发布消息
        turtle_vel_pub.publish(vel_msg);
        //void ros::Publisher::publish(const M & message) const:发布与此 Publisher 关联的主题上的消息。
        //ros::Publisher::publish ros::Publisher类的一个成员函数，用于将消息发送到与该发布者关联的话题。
        //M是一个模板类型，表示消息的类型。M &message 表示函数参数 message 是消息的引用，& 表示这是一个引用，避免了消息的复制，提高了效率。
        //const 表示该参数是常量引用，意味着 publish() 函数内部不能修改传入的消息内容。
        ROS_INFO("Publish turtle velocity command[%0.2f m/s, %0.2f rad/s]",
                    vel_msg.linear.x, vel_msg.angular.z);

        // 设置循环频率延时
        loop_rate.sleep();
    }

    return 0;  // 程序正常结束
}
